src/rgw/rgw_lua_utils.h

```diff
@@ -124,41 +125,117 @@ void create_metatable(lua_State* L, bool toplevel, Upvalues... upvalues)
     // give table a name (in cae of "toplevel")
     lua_setglobal(L, MetaTable::TableName().c_str());
   }
+
   // create metatable
   [[maybe_unused]] const auto rc = luaL_newmetatable(L, MetaTable::Name().c_str());
+  const auto table_stack_pos = lua_gettop(L);
+  
+  // add "index" closure to metatable
   lua_pushliteral(L, "__index");
   for (const auto upvalue : upvalue_arr) {
     lua_pushlightuserdata(L, upvalue);
   }
   lua_pushcclosure(L, MetaTable::IndexClosure, upvals_size);
-  lua_rawset(L, -3);
+  lua_rawset(L, table_stack_pos);
+
+  // add "newindex" closure to metatable
   lua_pushliteral(L, "__newindex");
   for (const auto upvalue : upvalue_arr) {
     lua_pushlightuserdata(L, upvalue);
   }
   lua_pushcclosure(L, MetaTable::NewIndexClosure, upvals_size);
-  lua_rawset(L, -3);
+  lua_rawset(L, table_stack_pos);
+
+  // add "pairs" closure to metatable
   lua_pushliteral(L, "__pairs");
   for (const auto upvalue : upvalue_arr) {
     lua_pushlightuserdata(L, upvalue);
   }
   lua_pushcclosure(L, MetaTable::PairsClosure, upvals_size);
-  lua_rawset(L, -3);
+  lua_rawset(L, table_stack_pos);
+  
+  // add "len" closure to metatable
   lua_pushliteral(L, "__len");
   for (const auto upvalue : upvalue_arr) {
     lua_pushlightuserdata(L, upvalue);
   }
   lua_pushcclosure(L, MetaTable::LenClosure, upvals_size);
-  lua_rawset(L, -3);
+  lua_rawset(L, table_stack_pos);
+
   // tie metatable and table
+  ceph_assert(lua_gettop(L) == table_stack_pos);
+  lua_setmetatable(L, -2);
+}
+
+template<typename Type>
+int value_dtor(lua_State* L) {
+  auto value  = reinterpret_cast<Type*>(lua_touserdata(L, -1));
+  value->~Type();
+  return NO_RETURNVAL;
+}
+
+template<typename MetaTable, typename Type>
+void create_owning_metatable(lua_State* L, bool toplevel, Type&& value)
+{
+  const auto table_name = MetaTable::TableName() + typeid(Type).name();
+  // create lua owned user data
+  auto buff = lua_newuserdata(L, sizeof(Type));
+  new (buff) Type(std::move(value));
+  const auto userdata_stack_pos = lua_gettop(L);
+  if (toplevel) {
+    // duplicate the userdata to make sure it remain in the stack
+    lua_pushvalue(L, -1);
+    // give table a name (in cae of "toplevel")
+    lua_setglobal(L, table_name.c_str());
+  }
+
+  // create metatable
+  [[maybe_unused]] auto rc = luaL_newmetatable(L, table_name.c_str());
```

Comment:
Would it indicate a bug if the metatable already exists? I wonder if we can ceph_assert(rc != 0);

Reply:
this is actually quite common that the metatale already exist. for example:
```lua
print(Request.GenericAttributes["hello"])
print(Request.GenericAttributes["foo"])
```
will create the "GenericAttributesMeta" metatable passing "hello" as an upvalue to its closures, and then try to re-create it, passing "foo" as an upvalue to its closures.
so the same metatable is used twice, with different upvalues to the closures.

according to the unit tests this works, and there is no override or reordering issue with the upvalues. however, do you think we should make the metatables unique?

Reply:
My main concern is that you're setting these metatable fields (`__index` etc.) every time you call `create_owning_metatable`. That's kinda weird especially when the closures have state in the form of upvalues. It would look more normal if it were:

```
if (luaL_newmetatable(...)) {
  lua_pushliteral(L, "__gc");
  lua_pushX();
  lua_rawset(L, table_stack_pos);
  // etc.
}
```

Reply:
i looked into this. your suggestion would work in most cases, but it does not work when we have an iterable object that contains tables:
for example: `Request.UserAcl.Grants["something"]` is a `Grant`, which is a table with subfields.
and we need to have different metatables for:
```
Request.UserAcl.Grants["something"]
```
and:
```
Request.UserAcl.Grants["something else"]
```

i will create a separate PR for that, since this is not specific to the "owning" metatable.

Reply:
see: https://tracker.ceph.com/issues/58412

Reply:
I'm not sure I understand:

> ... and we need to have different metatables for: ...

I've never seen a reason to update the fields of a metadata after getting set/created. So this looks like a code smell to me.

Are you saying 58412 is the result of this? It would make sense to me but I don't know the object/metadata structure of this API so I can't say.

Reply:
you are right, this code is a bug. if we keep it, the second call will override the upvalues of the first call.
if we change it, to bail out of the function if the metatable exist, the upvalues of the second invocation will be ignored.
the correct fix is to have unique names for the metatables.
for example, for `Request.UserAcl.Grants["something"]` the name of the metatable should be:
`Request.UserAcl.Grants.something.Grant` and not just `Grant`
once this is done, we can bail ot of the function if the metatable exists (`rc == 0`) since we are sure it is the same one with identical upvalues